// xslt-processor/xpath/node-tests/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NodeTestText = exports.NodeTestPI = exports.NodeTestNC = exports.NodeTestName = exports.NodeTestElementOrAttribute = exports.NodeTestComment = exports.NodeTestAny = void 0;
var node_test_any_1 = require("./node-test-any");
Object.defineProperty(exports, "NodeTestAny", { enumerable: true, get: function () { return node_test_any_1.NodeTestAny; } });
var node_test_comment_1 = require("./node-test-comment");
Object.defineProperty(exports, "NodeTestComment", { enumerable: true, get: function () { return node_test_comment_1.NodeTestComment; } });
var node_test_element_or_attribute_1 = require("./node-test-element-or-attribute");
Object.defineProperty(exports, "NodeTestElementOrAttribute", { enumerable: true, get: function () { return node_test_element_or_attribute_1.NodeTestElementOrAttribute; } });
var node_test_name_1 = require("./node-test-name");
Object.defineProperty(exports, "NodeTestName", { enumerable: true, get: function () { return node_test_name_1.NodeTestName; } });
var node_test_nc_1 = require("./node-test-nc");
Object.defineProperty(exports, "NodeTestNC", { enumerable: true, get: function () { return node_test_nc_1.NodeTestNC; } });
var node_test_pi_1 = require("./node-test-pi");
Object.defineProperty(exports, "NodeTestPI", { enumerable: true, get: function () { return node_test_pi_1.NodeTestPI; } });
var node_test_text_1 = require("./node-test-text");
Object.defineProperty(exports, "NodeTestText", { enumerable: true, get: function () { return node_test_text_1.NodeTestText; } });
