// xslt-processor/xpath/xpath-grammar-rules.js
"use strict";
// All the nonterminals of the grammar. The nonterminal objects are
// identified by object identity; the labels are used in the debug
Object.defineProperty(exports, "__esModule", { value: true });
exports.XPathDigits = exports.XPathFilterExpr = exports.XPathUnionExpr = exports.XPathPathExpr = exports.XPathArgumentRemainder = exports.XPathFunctionCall = exports.XPathNumber = exports.XPathVariableReference = exports.XPathPrimaryExpr = exports.XPathExpr = exports.XPathLiteral = exports.XPathPredicate = exports.XPathNodeTest = exports.XPathStep = exports.XPathAbsoluteLocationPath = exports.XPathRelativeLocationPath = exports.XPathLocationPath = void 0;
// output only.
exports.XPathLocationPath = {
    label: 'LocationPath',
    key: undefined
};
exports.XPathRelativeLocationPath = {
    label: 'RelativeLocationPath',
    key: undefined
};
exports.XPathAbsoluteLocationPath = {
    label: 'AbsoluteLocationPath',
    key: undefined
};
exports.XPathStep = {
    label: 'Step',
    key: undefined
};
exports.XPathNodeTest = {
    label: 'NodeTest',
    key: undefined
};
exports.XPathPredicate = {
    label: 'Predicate',
    key: undefined
};
exports.XPathLiteral = {
    label: 'Literal',
    key: undefined
};
exports.XPathExpr = {
    label: 'Expr',
    key: undefined
};
exports.XPathPrimaryExpr = {
    label: 'PrimaryExpr',
    key: undefined
};
exports.XPathVariableReference = {
    label: 'Variablereference',
    key: undefined
};
exports.XPathNumber = {
    label: 'Number',
    key: undefined
};
exports.XPathFunctionCall = {
    label: 'FunctionCall',
    key: undefined
};
exports.XPathArgumentRemainder = {
    label: 'ArgumentRemainder',
    key: undefined
};
exports.XPathPathExpr = {
    label: 'PathExpr',
    key: undefined
};
exports.XPathUnionExpr = {
    label: 'UnionExpr',
    key: undefined
};
exports.XPathFilterExpr = {
    label: 'FilterExpr',
    key: undefined
};
exports.XPathDigits = {
    label: 'Digits',
    key: undefined
};
