// xslt-processor/dom/functions.js
"use strict";
// Copyright 2023-2024 Design Liquido
// Copyright 2018 Johannes Wilm
// Copyright 2005 Google Inc.
// All Rights Reserved
Object.defineProperty(exports, "__esModule", { value: true });
exports.domGetAttributeValue = domGetAttributeValue;
exports.domSetAttribute = domSetAttribute;
exports.domSetTransformedAttribute = domSetTransformedAttribute;
exports.domAppendChild = domAppendChild;
exports.domAppendTransformedChild = domAppendTransformedChild;
exports.domCreateTextNode = domCreateTextNode;
exports.domCreateTransformedTextNode = domCreateTransformedTextNode;
exports.domCreateElement = domCreateElement;
exports.domCreateCDATASection = domCreateCDATASection;
exports.domCreateComment = domCreateComment;
exports.domCreateDocumentFragment = domCreateDocumentFragment;
exports.domCreateDTDSection = domCreateDTDSection;
// Wrapper around DOM methods so we can condense their invocations.
function domGetAttributeValue(node, name) {
    return node.getAttributeValue(name);
}
function domSetAttribute(node, name, value) {
    return node.setAttribute(name, value);
}
function domSetTransformedAttribute(node, name, value) {
    return node.setTransformedAttribute(name, value);
}
function domAppendChild(node, child) {
    return node.appendChild(child);
}
function domAppendTransformedChild(node, child) {
    return node.appendTransformedChild(child);
}
function domCreateTextNode(node, text) {
    return node.createTextNode(text);
}
function domCreateTransformedTextNode(node, text) {
    return node.createTransformedTextNode(text);
}
function domCreateElement(doc, name) {
    return doc.createElement(name);
}
function domCreateCDATASection(doc, data) {
    return doc.createCDATASection(data);
}
function domCreateComment(doc, text) {
    return doc.createComment(text);
}
function domCreateDocumentFragment(doc) {
    return doc.createDocumentFragment();
}
function domCreateDTDSection(doc, data) {
    return doc.createDTDSection(data);
}
//XDocument.prototype = new XNode(DOM_DOCUMENT_NODE, '#document');
